src/lem/circuit.rs

```diff
@@ -0,0 +1,904 @@
+}
+
+impl<F: LurkField> GlobalAllocator<F> {
+    /// Checks if the allocation for a numeric variable has already been cached.
+    /// If so, return the cached allocation variable. Allocate, cache and return
+    /// otherwise.
+    pub(crate) fn get_or_alloc_num<CS: ConstraintSystem<F>>(
+        &mut self,
+        cs: &mut CS,
+        f: F,
+    ) -> Result<AllocatedNum<F>> {
+        let wrap = FWrap(f);
+        match self.0.get(&wrap) {
+            Some(allocated_num) => Ok(allocated_num.to_owned()),
+            None => {
+                let allocated_num =
+                    allocate_num(cs, &format!("globally allocate {}", f.hex_digits()), f)?;
+                self.0.insert(wrap, allocated_num.clone());
+                Ok(allocated_num)
+            }
+        }
+    }
+}
+
+#[derive(Clone, Copy, PartialEq, Eq, Hash)]
+pub(crate) enum SlotType {
+    Hash2,
+    Hash3,
+    Hash4,
+}
+
+impl SlotType {
+    pub(crate) fn preimg_size(&self) -> usize {
+        match self {
+            Self::Hash2 => 4,
+            Self::Hash3 => 6,
+            Self::Hash4 => 8,
+        }
+    }
+}
+
+impl std::fmt::Display for SlotType {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::Hash2 => write!(f, "Hash2"),
+            Self::Hash3 => write!(f, "Hash3"),
+            Self::Hash4 => write!(f, "Hash4"),
+        }
+    }
+}
+
+/// A `Slot` is characterized by an index and a type
+#[derive(Clone, Copy, PartialEq, Eq, Hash)]
+struct Slot {
+    idx: usize,
+    typ: SlotType,
+}
+
+impl std::fmt::Display for Slot {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "Slot({}, {})", self.idx, self.typ)
+    }
+}
+
+impl LEM {
+    fn allocate_ptr<F: LurkField, CS: ConstraintSystem<F>>(
+        cs: &mut CS,
+        z_ptr: &ZPtr<F>,
+        name: &AString,
+        allocated_ptrs: &mut HashMap<AString, AllocatedPtr<F>>,
+    ) -> Result<AllocatedPtr<F>> {
+        if allocated_ptrs.contains_key(name) {
+            bail!("{} already allocated", name);
+        };
+        let allocated_tag =
+            allocate_num(cs, &format!("allocate {name}'s tag"), z_ptr.tag.to_field())?;
+        let allocated_hash = allocate_num(cs, &format!("allocate {name}'s hash"), z_ptr.hash)?;
+        let allocated_ptr = AllocatedPtr::from_parts(allocated_tag, allocated_hash);
+        allocated_ptrs.insert(name.clone(), allocated_ptr.clone());
+        Ok(allocated_ptr)
+    }
+
+    fn allocate_input<F: LurkField, CS: ConstraintSystem<F>>(
+        &self,
+        cs: &mut CS,
+        store: &mut Store<F>,
+        frame: &Frame<F>,
+        allocated_ptrs: &mut HashMap<AString, AllocatedPtr<F>>,
+    ) -> Result<()> {
+        for (i, ptr) in frame.input.iter().enumerate() {
+            let name = &self.input_vars[i];
+            Self::allocate_ptr(cs, &store.hash_ptr(ptr)?, name, allocated_ptrs)?;
```

Comment:
This appears to be unconstrained. The current treatment of public inputs is vague in this work, so it's not clear how the input is meant to be treated. I think this needs to be formalized somehow — either by implementing `Circuit`, `StepCircuit`, or in some other very explicit way. Otherwise the ambiguity of these unconstrained inputs (and outputs, I guess) is a bomb waiting to go off. I'll try to satisfy myself, but without a clear explanation of what is intended, implicit knowledge of what has been left undone may slip through the cracks.

Reply:
Inputs are unconstrained unless they are used somewhere within LEM. Say, for instance, you right a LEM that simply returns the number 0. Then that would mean this is a constant function, that ignores the input. The input should just be properly unconstrained. When it comes to output, they should always be constrained, but this is ensured by the fact that LEM always returns, and a return will constrain the output

Reply:
I am also confused how input is going to work. Initially I thought it would be using `inputize`, because that was how it was implemented.

Reply:
Inside a multiframe we have that next input must be equal to previous output. Between multiframes we must implement something like `StepCircuit`.